src/onnx/parse_matmul.cpp

```diff
@@ -38,6 +38,77 @@ struct parse_matmul : op_parser<parse_matmul>
+                                     instruction_ref& ba0,
+                                     instruction_ref& ba1)
+    {
+        // try broadcasting if dimensions other than last two do not match
+        if(not std::equal(
+               s0_lens.rbegin() + 2, s0_lens.rend(), s1_lens.rbegin() + 2, s1_lens.rend()))
+        {
+            auto l0_it = s0_lens.begin() + s0_lens.size() - 2;
+            std::vector<std::size_t> l0_broadcasted_lens(s0_lens.begin(), l0_it);
+            auto l1_it = s1_lens.begin() + s1_lens.size() - 2;
+            std::vector<std::size_t> l1_broadcasted_lens(s1_lens.begin(), l1_it);
+            auto output_lens = compute_broadcasted_lens(l0_broadcasted_lens, l1_broadcasted_lens);
+            l0_broadcasted_lens = output_lens;
+            l0_broadcasted_lens.insert(l0_broadcasted_lens.end(), l0_it, s0_lens.end());
+            l1_broadcasted_lens = output_lens;
+            l1_broadcasted_lens.insert(l1_broadcasted_lens.end(), l1_it, s1_lens.end());
+            if(s0_lens != l0_broadcasted_lens)
+            {
+                ba0 = info.add_instruction(
+                    make_op("multibroadcast", {{"out_lens", l0_broadcasted_lens}}), a0);
+            }
+            if(s1_lens != l1_broadcasted_lens)
+            {
+                ba1 = info.add_instruction(
+                    make_op("multibroadcast", {{"out_lens", l1_broadcasted_lens}}), a1);
+            }
+        }
+    }
+
+    // Convert to int16 prior to a shift to ensure we preserve accuracy here then
+    // convert back to int8
+    static instruction_ref add_int8_shift(const onnx_parser::node_info& info,
+                                          instruction_ref& unshifted_input)
+    {
+        auto int8_shift = info.add_literal(
+            migraphx::literal{migraphx::shape{migraphx::shape::int16_type}, {-128}});
+
+        auto unshifted_input_int16 = info.add_instruction(
+            migraphx::make_op("convert", {{"target_type", migraphx::shape::int16_type}}),
+            unshifted_input);
+
+        auto input_shifted_int16 = info.add_common_op("add", unshifted_input_int16, int8_shift);
+
+        return info.add_instruction(
+            migraphx::make_op("convert", {{"target_type", migraphx::shape::int8_type}}),
+            input_shifted_int16);
+    }
+
+    static instruction_ref set_bias_arg(const onnx_parser::node_info& info,
+                                        const std::vector<instruction_ref>& args,
+                                        const int index,
+                                        const instruction_ref& input)
+    {
+        if(args.size() > index)
+        {
+            instruction_ref bias_arg = args[index];
+            if(bias_arg->get_shape().type() != input->get_shape().type())
+            {
+                MIGRAPHX_THROW("PARSE_QUANT_DOT: zero point must be the same type as data");
+            }
+
+            return info.add_common_op("sub", input, bias_arg);
```

Comment:
This would do "sub" between two uint8_t types which would result in incorrect results. 
It should do subtraction in signed type

Reply:
https://onnx.ai/onnx/operators/onnx__MatMulInteger.html#matmulinteger

A and AZ will have same type and B and BZ will have same type. 

For the `matmulInteger(A-AZ, B-ZA)`  we can have four cases. 

1. `A : int8` `B : uint8` --> in this case we would need to convert entire operation to float type. Because for the `B - BZ` in worse case can generate `-255` as the value and that would require GEMM to happen in float16/32  to represent all values properly. 
2. `A: int8`, `B:int8` --> we already handle that case but `B - BZ` or `A - AZ` can again give `-128 - 128` = `-255` value which can't be handled. Therefore entire GEMM need to happen in float16/32
3. `A: uint8` and `B:uint8` same case as 2. it may generate `-255` 
4. `A:uint8` and  `B:int8` same as case 1. 
5. 
Here is the reference python implementation from onnx : 
https://github.com/onnx/onnx/blob/main/onnx/reference/ops/op_matmul_integer.py#L15

I think what is required is 
1. If any input and arg and corresponding zero point is `uint8` then first convert them to `int8` by adding `-128` shift. 
2. then compute `A - AZ` and `B-BZ`
3. And then do the `int8 * int8 = int32` 

Right now what you are doing is following : 
1.  compute `A-AZ` and `B-BZ` which would suffer from unsigned subtraction issue. Ideally we should throw warning for unsigned subtraction. 
2.  If any of the input is uint8 and then convert to int8 by shifting -128. 
3.  And then do `int8 * int8 = int32`

Reply:
Done